platform/linux-generic/odp_packet.c

```diff
@@ -2468,6 +2473,97 @@ int _odp_packet_ipv4_chksum_insert(odp_packet_t pkt)
 					2, &chksum);
 }
 
+static int _odp_packet_tcp_udp_chksum_insert(odp_packet_t pkt, uint16_t proto)
+{
+	odp_packet_hdr_t *pkt_hdr = packet_hdr(pkt);
+	uint32_t zero = 0;
+	uint32_t sum;
+	uint16_t l3_ver;
+	uint16_t chksum;
+	uint32_t chksum_offset;
+
+	if (pkt_hdr->p.l3_offset == ODP_PACKET_OFFSET_INVALID)
+		return -1;
+	if (pkt_hdr->p.l4_offset == ODP_PACKET_OFFSET_INVALID)
+		return -1;
+
+	odp_packet_copy_to_mem(pkt, pkt_hdr->p.l3_offset, 2, &l3_ver);
+
+	if (_ODP_IPV4HDR_VER(l3_ver) == _ODP_IPV4)
+		sum = packet_sum_ones_comp16_32(pkt_hdr,
+						pkt_hdr->p.l3_offset +
+						_ODP_IPV4ADDR_OFFSSET,
+						2 * _ODP_IPV4ADDR_LEN,
+						false);
+	else
+		sum = packet_sum_ones_comp16_32(pkt_hdr,
+						pkt_hdr->p.l3_offset +
+						_ODP_IPV6ADDR_OFFSSET,
+						2 * _ODP_IPV6ADDR_LEN,
+						false);
+#if ODP_BYTE_ORDER == ODP_BIG_ENDIAN
+	sum += proto;
+#else
+	sum += proto << 8;
+#endif
+
+	if (proto == _ODP_IPPROTO_TCP) {
+		sum += _odp_cpu_to_be_16(pkt_hdr->frame_len -
+					 pkt_hdr->p.l4_offset);
+		chksum_offset = pkt_hdr->p.l4_offset + _ODP_UDP_CSUM_OFFSET;
+	} else {
+		sum += packet_sum_ones_comp16_32(pkt_hdr,
+						 pkt_hdr->p.l4_offset +
+						 _ODP_UDP_LEN_OFFSET,
+						 2,
+						 false);
+		chksum_offset = pkt_hdr->p.l4_offset + _ODP_UDP_CSUM_OFFSET;
+	}
+	odp_packet_copy_from_mem(pkt, chksum_offset, 2, &zero);
+
+	sum += packet_sum_ones_comp16_32(pkt_hdr,
+					 pkt_hdr->p.l4_offset,
+					 pkt_hdr->frame_len -
+					 pkt_hdr->p.l4_offset,
+					 false);
+
+	/* Not more than two additions */
+	sum = (sum & 0xffff) + (sum >> 16);
+	sum = (sum & 0xffff) + (sum >> 16);
+
+	chksum = ~sum;
+
+	return odp_packet_copy_from_mem(pkt,
+					chksum_offset,
+					2, &chksum);
+}
+
+/**
+ * Calculate and fill in TCP checksum
+ *
+ * @param pkt  ODP packet
+ *
+ * @retval 0 on success
+ * @retval <0 on failure
+ */
+int _odp_packet_tcp_chksum_insert(odp_packet_t pkt)
+{
+	return _odp_packet_tcp_udp_chksum_insert(pkt, _ODP_IPPROTO_TCP);
+}
+
+/**
+ * Calculate and fill in UDP checksum
+ *
+ * @param pkt  ODP packet
+ *
+ * @retval 0 on success
+ * @retval <0 on failure
+ */
+int _odp_packet_udp_chksum_insert(odp_packet_t pkt)
+{
+	return _odp_packet_tcp_udp_chksum_insert(pkt, _ODP_IPPROTO_UDP);
```

Comment:
This code does not seem to handle the case when the calculated UDP checksum is 0 in which case it should be transmitted as 0xffff.

Reply:
@JannePeltonen good catch, thank you. Should be fixed now.